File: app/src/main/java/de/kontranik/freebudget/adapter/TransactionAdapter.java
package de.kontranik.freebudget.adapter;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.List;
import java.util.Locale;

import de.kontranik.freebudget.R;
import de.kontranik.freebudget.config.Config;
import de.kontranik.freebudget.fragment.OverviewFragment;
import de.kontranik.freebudget.model.Transaction;

import static de.kontranik.freebudget.config.Config.DATE_SHORT;

public class TransactionAdapter extends ArrayAdapter<Transaction>  {

    private LayoutInflater inflater;
    private int layout;
    private List<Transaction> transactions;
    private boolean markLastEdited;

    public TransactionAdapter(Context context, int resource, List<Transaction> transactions) {
        super(context, resource, transactions);
        this.transactions = transactions;
        this.layout = resource;
        this.inflater = LayoutInflater.from(context);

        SharedPreferences settings = context.getSharedPreferences(Config.PREFS_FILE, Context.MODE_PRIVATE);
        markLastEdited  = settings.getBoolean(Config.PREF_MARK_LAST_EDITED, false);
    }

    public View getView(int position, View convertView, ViewGroup parent) {

        ViewHolder viewHolder;
        if(convertView==null){
            convertView = inflater.inflate(this.layout, parent, false);
            viewHolder = new ViewHolder(convertView);
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }

        Transaction transaction = transactions.get(position);

        String amount_fact = String.format(Locale.getDefault(), "%1$,.2f", transaction.getAmount_fact());
        String amount_planned = String.format(Locale.getDefault(), "%1$,.2f",transaction.getAmount_planned());

        viewHolder.textView_amount_planned.setText(amount_planned);
        viewHolder.textView_amount_fact.setText(amount_fact);

        if (transaction.getAmount_fact() > 0 ) {
            viewHolder.textView_amount_fact.setTextColor(
                    ContextCompat.getColor(parent.getContext(), R.color.colorGreen));
         } else if ( transaction.getAmount_fact() < 0) {
            viewHolder.textView_amount_fact.setTextColor(
                    ContextCompat.getColor(parent.getContext(), R.color.colorRed));
        }

        if ( transaction.getAmount_planned() > 0) {
            viewHolder.textView_amount_planned.setTextColor(
                    ContextCompat.getColor(parent.getContext(), R.color.colorGreen));
        } else if ( transaction.getAmount_planned() < 0) {
            viewHolder.textView_amount_planned.setTextColor(
                    ContextCompat.getColor(parent.getContext(), R.color.colorRed));
        }

        viewHolder.descriptionView.setText(transaction.getDescription());

        DateFormat df = new SimpleDateFormat(DATE_SHORT, Locale.getDefault());
        String dateString = "not set";
        if ( transaction.getDate() > 0 ) {
            dateString = df.format(transaction.getDate());
        }

        viewHolder.categoryView.setText(
                parent.getResources().getString(
                        R.string.subTitleTransaction, dateString, transaction.getCategory()) );

        // last edited hervorheben
        if ( markLastEdited && transaction.getId() == OverviewFragment.lastEditedId ) {
            viewHolder.descriptionView.setTextColor( ContextCompat.getColor(parent.getContext(), R.color.colorBackgroundListItem) );
            viewHolder.textView_amount_planned.setTextColor( ContextCompat.getColor(parent.getContext(), R.color.colorBackgroundListItem) );
            viewHolder.textView_amount_fact.setTextColor( ContextCompat.getColor(parent.getContext(), R.color.colorBackgroundListItem) );
            viewHolder.categoryView.setTextColor( ContextCompat.getColor(parent.getContext(), R.color.colorBackgroundListItem) );
            viewHolder.linearLayout_Item.setBackgroundColor( ContextCompat.getColor(parent.getContext(), R.color.colorBackgroundAccent) );
        } else {
            viewHolder.descriptionView.setTextColor( ContextCompat.getColor(parent.getContext(), R.color.colorTextListItem) );
            viewHolder.categoryView.setTextColor( ContextCompat.getColor(parent.getContext(), R.color.colorTextListItem) );
            viewHolder.textView_amount_planned.setTextColor( ContextCompat.getColor(parent.getContext(), R.color.colorTextListItem) );
            viewHolder.textView_amount_fact.setTextColor( ContextCompat.getColor(parent.getContext(), R.color.colorTextListItem) );
            viewHolder.linearLayout_Item.setBackgroundColor( ContextCompat.getColor(parent.getContext(), R.color.colorBackgroundListItem ));
        }

        return convertView;
    }

    private class ViewHolder {
        final LinearLayout linearLayout_Item;
        final TextView textView_amount_planned, textView_amount_fact, descriptionView, categoryView;
        ViewHolder(View view) {
            linearLayout_Item = view.findViewById(R.id.linearLayout_Item);
            textView_amount_planned = view.findViewById(R.id.textView_amount_planned);
            textView_amount_fact = view.findViewById(R.id.textView_amount_fact);
            descriptionView = view.findViewById(R.id.textView_description);
            categoryView = view.findViewById(R.id.textView_category);
        }
    }
}

File: app/src/main/java/de/kontranik/freebudget/adapter/TransactionAdapter.kt
package de.kontranik.freebudget.adapter

import android.content.Context
import android.graphics.Color
import android.widget.ArrayAdapter
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import de.kontranik.freebudget.R
import de.kontranik.freebudget.fragment.AllTransactionFragment
import android.widget.LinearLayout
import android.widget.TextView
import android.view.View
import de.kontranik.freebudget.config.Config
import de.kontranik.freebudget.model.Transaction
import java.text.DateFormat
import java.util.*

class TransactionAdapter(
    context: Context,
    private val layout: Int,
    private val transactions: List<Transaction>
) : ArrayAdapter<Transaction?>(context, layout, transactions) {
    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private val markLastEdited: Boolean
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        val viewHolder: ViewHolder
        if (view == null) {
            view = inflater.inflate(layout, parent, false)
            viewHolder = ViewHolder(view)
            view.tag = viewHolder
        } else {
            viewHolder = view.tag as ViewHolder
        }
        val transaction = transactions[position]
        val amountFact = String.format(Locale.getDefault(), "%1$,.2f", transaction.amountFact)
        val amountPlanned =
            String.format(Locale.getDefault(), "%1$,.2f", transaction.amountPlanned)
        viewHolder.textviewAmountPlanned.text = amountPlanned
        viewHolder.textviewAmountFact.text = amountFact
        when {
            transaction.amountFact > 0 -> {
                viewHolder.textviewAmountFact.setTextColor(
                    ContextCompat.getColor(parent.context, R.color.colorGreen)
                )
            }
            transaction.amountFact < 0 -> {
                viewHolder.textviewAmountFact.setTextColor(
                    ContextCompat.getColor(parent.context, R.color.colorRed)
                )
            }
            else -> {
                viewHolder.textviewAmountFact.setTextColor(
                    ContextCompat.getColor(parent.context, R.color.colorTextListItem)
                )
            }
        }
        when {
            transaction.amountPlanned > 0 -> {
                viewHolder.textviewAmountPlanned.setTextColor(
                    ContextCompat.getColor(parent.context, R.color.colorGreen)
                )
            }
            transaction.amountPlanned < 0 -> {
                viewHolder.textviewAmountPlanned.setTextColor(
                    ContextCompat.getColor(parent.context, R.color.colorRed)
                )
            }
            else -> {
                viewHolder.textviewAmountPlanned.setTextColor(
                    ContextCompat.getColor(parent.context, R.color.colorTextListItem)
                )
            }
        }
        viewHolder.descriptionView.text = transaction.description
        val df = DateFormat.getDateInstance(DateFormat.MEDIUM, Locale.getDefault())
        var dateString: String? = parent.context.getString(R.string.not_setted)
        if (transaction.date > 0) {
            dateString = df.format(transaction.date)
        }
        viewHolder.categoryView.text = parent.resources.getString(
            R.string.subTitleTransaction, dateString, transaction.category
        )

        // last edited hervorheben
        if (markLastEdited && transaction.id == AllTransactionFragment.lastEditedId) {
            viewHolder.linearlayoutItem.setBackgroundColor(
                ContextCompat.getColor(
                    parent.context,
                    R.color.colorBackgroundAccent
                )
            )
        } else {
            viewHolder.linearlayoutItem.setBackgroundColor(Color.TRANSPARENT)
        }
        return view!!
    }

    private inner class ViewHolder internal constructor(view: View) {
        val linearlayoutItem: LinearLayout = view.findViewById(R.id.linearLayout_Item)
        val textviewAmountPlanned: TextView = view.findViewById(R.id.textView_amount_planned)
        val textviewAmountFact: TextView = view.findViewById(R.id.textView_amount_fact)
        val descriptionView: TextView = view.findViewById(R.id.textView_description)
        val categoryView: TextView = view.findViewById(R.id.textView_category)
    }

    init {
        val settings = context.getSharedPreferences(Config.PREFS_FILE, Context.MODE_PRIVATE)
        markLastEdited = settings.getBoolean(Config.PREF_MARK_LAST_EDITED, false)
    }
}